six/modules/c++/samples/test_large_offset.cpp

```diff
@@ -0,0 +1,150 @@
+const size_t ROWS_TO_SKIP = NUM_ROWS_IN_FIRST_SEGMENT - 1000;
+const size_t SICD_ELEMENT_SIZE = sizeof(std::complex<sys::Int16_T>);
+const size_t SIDD_ELEMENT_SIZE = sizeof(sys::Int16_T);
+
+void createNITF(const std::string& inputPathname,
+        const std::string& outputPathname)
+{
+    six::XMLControlRegistry registry;
+    registry.addCreator(six::DataType::COMPLEX,
+            new six::XMLControlCreatorT<
+                    six::sicd::ComplexXMLControl>());
+    registry.addCreator(six::DataType::DERIVED,
+            new six::XMLControlCreatorT<
+                    six::sidd::DerivedXMLControl>());
+    const std::vector<std::string> schemaPaths;
+    six::NITFReadControl reader;
+    reader.setXMLControlRegistry(&registry);
+    reader.load(inputPathname, schemaPaths);
+
+    std::auto_ptr<six::Data> data(reader.getContainer()->getData(0)->clone());
+    size_t elementSize;
+    if (data->getDataType() == six::DataType::COMPLEX)
+    {
+        data->setNumRows(52440); // (11GB / 4) ^ (1/2)
+        data->setNumCols(52440);
+        data->setPixelType(six::PixelType::RE16I_IM16I);
+        elementSize = SICD_ELEMENT_SIZE;
+    }
+    else
+    {
+        data->setNumRows(74152); // (11GB / 2) ^ (1/2)
+        data->setNumCols(74162);
+        data->setPixelType(six::PixelType::MONO16I);
+        elementSize = SIDD_ELEMENT_SIZE;
+    }
+
+    const size_t AREA = data->getNumRows() * data->getNumCols();
+    const size_t FIRST_SIZE = elementSize * data->getNumCols() * ROWS_TO_SKIP;
+    mem::SharedPtr<six::Container> container(
+            new six::Container(data->getDataType()));
+    container->addData(data.release());
+
+    six::NITFWriteControl writer;
+    writer.setXMLControlRegistry(&registry);
+    writer.initialize(container);
+
+    mem::ScopedArray<six::UByte> imageData(new six::UByte[elementSize * AREA]);
+    std::memset(imageData.get(), 0, FIRST_SIZE);
+    std::memset(imageData.get() + FIRST_SIZE, 1,
+            (AREA * elementSize) - FIRST_SIZE);
+
+    writer.save(imageData.get(), outputPathname);
+}
+
+bool checkNITF(const std::string& pathname)
+{
+    six::XMLControlRegistry registry;
+    registry.addCreator(six::DataType::COMPLEX,
+            new six::XMLControlCreatorT<
+                    six::sicd::ComplexXMLControl>());
+    registry.addCreator(six::DataType::DERIVED,
+            new six::XMLControlCreatorT<
+                    six::sidd::DerivedXMLControl>());
+
+    six::NITFReadControl reader;
+    reader.setXMLControlRegistry(&registry);
+    const std::vector<std::string> schemaPaths;
+    reader.load(pathname, schemaPaths);
+    std::auto_ptr<six::Data> data(reader.getContainer()->getData(0)->clone());
+
+    six::Region region;
+    region.setStartRow(ROWS_TO_SKIP);
+    region.setNumRows(data->getNumRows() - ROWS_TO_SKIP);
+    region.setStartCol(0);
+    reader.interleaved(region, 0);
+    return static_cast<sys::Int8_T>(region.getBuffer()[0]) == 1;
```

Comment:
So this test would have caught the old bug (which is good), but it's not as comprehensive as it could be.  Rather than just some memsets upfront, I think it would be a better test to fill in each pixel with incremental data (i.e. have a for loop that sets each pixel to the value of the for loop counter or something) and then here verify that the region that's read out matches what's expected for all pixel values rather than the first one.